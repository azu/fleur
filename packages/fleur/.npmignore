/src
/tmp
/jest.config.js
/tsconfig.json
/tsconfig.test.json
/webpack.config.js
/README.md
