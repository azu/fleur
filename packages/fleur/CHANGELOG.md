### 0.0.11
- Suport Redux Devtools Extension
  ```typescript
  import Fleur, { withReduxDevTools } from '@ragg/fleur'
  const app = new Fleur({ ... })
  const context = withReduxDevTools(app.createContext())
  ```

### 0.0.9
- Update dependency (immer -> 1.8.0)

### 0.0.8
#### Behavior changes
- `Store#emitChange` is batched by `requestAnimationFrame` only client side.
  (In server side, synced not batched)

### 0.0.7
#### API Changes
- Accept action name in first argument of `action()`
  ``` typescript
  const increase = action<{ amount: number }>('increase')
  ```
- Add `actions()` for auto naming actions
  ```typescript
  export const CountActions = actions('Count', {
      increase: action<{ amount: number }>(), // It names 'Count/increase'
  })
  ```
- `ExtractActionIdentifiersFromObject` renamed to `ActionsOf`

### 0.0.6
#### API Changes
- Allow to any types in Store#rehydrate,dehydrate

### 0.0.5
#### Fixes
- Fix missing async-process waiting for OperationContext#e
xecuteOperation

### 0.0.4
#### API Changes
- Accept storeName: string in AppContext.getStore for improve debuggerbility
  No supports in another context getStore methods. (Usable only AppContext.getStore)

### 0.0.3
#### API Changes
- Accept many listener per Action in one Store
  ```ts
  // invalid in 0.0.2
  class SomeStore extends Store {
      private handleSomeAction = listen(someAction, () => { /* Do something */ })
      private handleSomeAction2 = listen(someAction, () => { /* Do something */ }) // error!
  }

  // Can in 0.0.3
  class SomeStore extends Store {
      private handleSomeAction = listen(someAction, () => { /* Do something */ })
      private handleSomeAction2 = listen(someAction, () => { /* Do something */ })
  }
  ```

### 0.0.2
#### API Changes
- Start changelog.
- Rename `Store#produce` to `Store#updateWith`.
- `static storeName = 'someStoreName'` is required for Store.

#### Fixes
- Broken store rehydration state under the Uglify
