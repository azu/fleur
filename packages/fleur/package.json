{
  "name": "@ragg/fleur",
  "version": "0.0.12",
  "homepage": "https://github.com/ra-gg/fleur/tree/master/packages/fleur",
  "main": "lib/index.js",
  "types": "typings/index.d.ts",
  "author": "Ragg<ragg.devpr@gmail.com>",
  "license": "MIT",
  "scripts": {
    "prepublishOnly": "rm -rf ./{typings,lib} && tsc --emitDeclarationOnly --declaration --outDir ./typings && tsc --outDir lib",
    "test": "jest --verbose"
  },
  "devDependencies": {
    "@types/eventemitter3": "^2.0.2",
    "@types/flux-standard-action": "1.1.0",
    "@types/invariant": "2.2.29",
    "cache-loader": "1.2.2",
    "eventemitter3": "3.1.0",
    "jest": "24.8.0",
    "ts-jest": "23.10.4",
    "typescript": "3.4.5"
  },
  "dependencies": {
    "immer": "1.8.0",
    "invariant": "2.2.4"
  }
}
