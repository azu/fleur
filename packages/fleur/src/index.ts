export { ActionIdentifier } from './Action'
export { Fleur as default } from './Fleur' // Aliasing for editor intellisense
export { Store, StoreClass, listen } from './Store'
export { StoreContext } from './StoreContext'
export { OperationContext } from './OperationContext'
export { action, actions, ActionsOf } from './Action'
export { AppContext } from './AppContext'
export { ComponentContext } from './ComponentContext'
export { operations, operation } from './Operations'
export { withReduxDevTools } from './withReduxDevtools'
