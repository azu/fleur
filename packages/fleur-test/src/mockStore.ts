import { StoreClass, Store } from '@ragg/fleur'

type ExtractStateType<T extends Store> = T extends Store<infer S> ? S : never

export interface MockStore {
    name: string
    store: Store<any>
}

export const mockStore = <S extends StoreClass>(StoreClass: S, partialState: Partial<ExtractStateType<InstanceType<S>>>): MockStore => {
    const store = new StoreClass()
    Object.assign((store as any).state, partialState)
    return { name: StoreClass.storeName, store }
}
