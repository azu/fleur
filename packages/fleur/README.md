# ðŸŒ¼ Fleur ðŸŒ¼ [![npm version](https://badge.fury.io/js/%40ragg%2Ffleur.svg)](https://www.npmjs.com/package/@ragg/fleur) [![travis](https://travis-ci.org/ra-gg/fleur.svg?branch=master)](https://travis-ci.org/ra-gg/fleur)
An Fully-typed Flux framework inspired by Fluxible.
Runs on Node / Web.

(No dependence to React. see [this](https://www.npmjs.com/package/@ragg/fleur-react) if you want to use with React.)

## Example

``` typescript
// actions.ts (Action typings)
import { action } from '@ragg/fleur'

export const increase = action<{ amount: number }>();
export const decrease = action<{ amount: number }>();
```

``` typescript
// store.ts (Store)
import { listen, Store } from '@ragg/fleur'
import * as actions from './actions.ts'

export default class SomeStore extends Store {
    public state: { count: number } = { count: 0 }

    private handleIncrease = listen(actions.increase, (payload) => {
        // `this.updateWith` is immutable changing `this.state` with `immer.js`
        this.updateWith(draft => draft.count += payload.amount)
    })

    private handleDecrease = listen(actions.decrease, (payload) => {
        this.updateWith(draft => draft.count -= payload.amount)
    })

    public getCount() {
        return this.state.count
    }
}
```

``` typescript
// operations.ts (Action Creator)
import { operation } from '@ragg/fleur'
import * as actions from './actions.ts'

export const increaseOperation = operation((ctx, { amount }: { amount: number }) => {
    ctx.dispatch(actions.increase, { amount })
})

export const decreaseOperation = operation((ctx, { amount }: { amount: number }) => {
    ctx.dispatch(actions.decrease, { amount })
})
```

``` typescript
// app.ts
import Fleur from '@ragg/fleur'
import SomeStore from './store.ts'

const app = new Fleur({
    stores: [SomeStore],
})

const ctx = app.createContext()
ctx.executeOperation(increaseOperation, { increase: 10 })
console.log(ctx.getStore(SomeStore).getCount()) // => 10
```

## How to use with React?
See [`@ragg/fleur-react`](https://www.npmjs.com/package/@ragg/fleur-react).
