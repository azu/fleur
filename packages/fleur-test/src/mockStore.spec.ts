import { mockStore } from './mockStore'
import { Store } from '@ragg/fleur'

describe('mockStoreContext', () => {
  it('Should create Store entry correctly', () => {
    class SomeStore extends Store<{ state: string }> {
      static storeName = 'SomeStore'
      public state = { state: '' }
    }

    const actual = mockStore(SomeStore, { state: 'mock' })
    expect(actual).toMatchSnapshot()
  })
})
