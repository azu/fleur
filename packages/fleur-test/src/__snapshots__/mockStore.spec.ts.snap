// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mockStoreContext Should create Store entry correctly 1`] = `
Object {
  "name": "SomeStore",
  "store": SomeStore {
    "listeners": Object {},
    "requestId": null,
    "state": Object {
      "state": "mock",
    },
  },
}
`;
